Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,222 @@
+      ()
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Client is closed, cannot send a message")
+    }
+  }
+  
+  mutating func send(status: String, trailingMetadata: Metadata) throws {
+    // Only server does this.
+  }
+  
+  /// Returns the client's next request to the server.
+  /// - Returns: The request to be made to the server.
+  mutating func nextRequest() throws -> ByteBuffer? {
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Client is not open yet.")
+    case .clientOpenServerIdle(var clientOpenState):
+      let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+      self.state = .clientOpenServerIdle(clientOpenState)
+      return request
+    case .clientOpenServerOpen(var clientOpenState):
+      let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+      self.state = .clientOpenServerOpen(clientOpenState)
+      return request
+    case .clientOpenServerClosed:
+      // Nothing to do: no point in sending request if server is closed.
+      return nil
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Can't send request if client is closed")
+    }
+  }
+  
+  mutating func receive(metadata: Metadata, endStream: Bool) {
+    // This is metadata received by the client from the server.
+    // It can be initial, which confirms that the server is now open;
+    // or an END_STREAM trailer, meaning the response is over.
+    if endStream {
+      self.clientReceivedEndHeader()
+    } else {
+      self.clientReceivedMetadata()
+    }
+  }
+  
+  mutating func receive(message: ByteBuffer, messageProcessor: ([UInt8]) throws -> Void) throws {
+    // This is a message received by the client, from the server.
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Cannot have received anything from server if client is not yet open.")
+    case .clientOpenServerIdle(let clientOpenState):
+      preconditionFailure("Server cannot have sent a message before sending the initial metadata.")
+    case .clientOpenServerOpen(let clientOpenState):
+      try clientOpenState.deframer.process(buffer: message, messageProcessor)
+    case .clientOpenServerClosed:
+      preconditionFailure("Cannot have received anything from a closed server.")
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Shouldn't receive anything if client's closed.")
+    }
+  }
+  
+  // - MARK: Client-opening transitions
+
+  mutating func clientSentMetadata() throws {
+    // Client sends metadata only when opening the stream.
+    // They send grpc-timeout and method name along with it.
+    // TODO: should these things be validated in the handler or here?
```

Comment:
For the timeout we'll insert it at a higher level (as it isn't transport specific) so no need to verify it here. For the method name (which is transport specific) we should insert it here, which I think means the state machine will need to hold the `MethodDescriptor` of the RPC we're going to call.